pypiserver/__main__.py

```diff
@@ -36,14 +42,132 @@ def init_logging(
+            self.log_date_time_string(),
+            format % args,
+        )
+
+
+class ServerCheck:
+    """Methods to check for a server's presence in the import path."""
+
+    # Because we need to potentially override behavior for different servers
+    # by passing options into bottle's `run` method, and because those options
+    # aren't necessarily compatible between servers, we need to be able to guess
+    # which server bottle will pick if the user specifies the `auto` server
+    # type. These methods' names match the `__name__` property of the server
+    # adapters in bottle, so we can filter the adapters on the `AutoServer`
+    # adapter to figure out which ones are present. Since bottle uses the
+    # first server to successfully import from that same list, this should
+    # be sufficient to give us good certainty as to which server it will run.
+
+    # pylint: disable=missing-function-docstring
+    # pylint: disable=invalid-name
+    # pylint: disable=import-outside-toplevel
+    # pylint: disable=unused-import
+
+    @staticmethod
+    def WaitressServer() -> bool:
+        try:
+            import waitress  # noqa: F401
+
+            return True
+        except ImportError:
+            return False
+
+    @staticmethod
+    def PasteServer() -> bool:
+        try:
+            import paste  # noqa: F401
+
+            return True
+        except ImportError:
+            return False
+
+    @staticmethod
+    def TwistedServer() -> bool:
+        try:
+            import twisted.web  # noqa: F401
+
+            return True
+        except ImportError:
+            return False
+
+    @staticmethod
+    def CherryPyServer() -> bool:
+        try:
+            import cheroot.wsgi  # noqa: F401
+
+            return True
+        except ImportError:
+            try:
+                import cherrypy.wsgiserver  # noqa: F401
+
+                return True
+            except ImportError:
+                return False
+
+    @staticmethod
+    def WSGIRefServer() -> bool:
+        try:
+            import wsgiref  # noqa: F401
+
+            return True
+        except ImportError:
+            return False
+
+    # pylint: enable=missing-function-docstring
+    # pylint: enable=invalid-name
+    # pylint: enable=import-outside-toplevel
+    # pylint: enable=unused-import
+
+
+def guess_auto_server() -> "t.Type[pypiserver.bottle.ServerAdapter]":
+    """Guess which server bottle will use for the auto setting."""
+    # pylint: disable=import-outside-toplevel
+    import pypiserver.bottle  # pylint: disable=redefined-outer-name
+
+    # Return the first ServerAdapter in `AutoServer.adapters` whose
+    # corresponding method in `ServerCheck` returns True. If none of the
+    # servers import correctly, `next()` will raise a `StopIteration` error.
+    return next(
+        filter(
+            lambda s: getattr(ServerCheck, s.__name__)(),
+            pypiserver.bottle.AutoServer.adapters,
+        )
+    )
```

Comment:
We're assuming a number of available servers here (the implemented methods of `ServerCheck` so we're not really decoupled from whatever is defined in `AutoServer.adapters`. Since we also use the result only for identification of a `server_method` (see my other comment below), we may just hard code a couple of `server_method` strings in here, which could simplify this function to something like this:

```python
def guess_auto_server():
    server_methods = [('waitress', 'waitress'), ('paste', 'paste'), ('twisted', 'twisted.web'), ('cherrypy', 'cherrypy.wsgiserver'), ('wsgiref','wsgiref')]
    return next(s[0] for s in server_methods if can_import(s[1]))
```

Also, do you think it's a good idea to slap a try/except StopIteration on it to give a more useful error message? I guess the chances of it failing with a StopIteration are slim because the `wsgiref` package should always be available, but idk, lettng StopIteration propagate seems kinda strange.

Reply:
The intent is to remain coupled to what's defined in `.adapters`, since that's bottle's source of truth on what can run. I'd like to be sure that if we pull in a bottle update that updates that list, we wind up with an error because we don't have a method for it in `ServerCheck`. 

We can catch the error and raise something more specific, for sure. My thinking was in line with yours, like it _should_ be impossible because `wsgiref` is in the standard library, but we might as well be safe.

Reply:
Hmm, if I understand you correctly, you are worried that we might inadvertently run out of sync in the way we resolve `auto` and `bottle` does. The way this function works now, in CI it goes through all items in `Autoserver.adapters` because we don't install any third party servers in CI so it ends up at `WSGIRefServer` always.  When a new `ServerAdapter` is added or a name changes, we do not have any corresponding method in `ServerCheck` and we get an error in CI (currently StopIteration), so that we know something has changed. Whenever the tests are run on another machine, behaviour is undefined since any third party servers might be installed.

This kind of sounds like a very implicit way of testing whether something has changed in bottle. Might be better to write an explicit test for this, if this is what you're worried about. When using strings, this would be something like reverse-mapping `bottle.server_names` and checking our list of strings against whatever `AutoServer.adapters` maps to.

Reply:
Yep, you understand my concern perfectly. We vendor bottle, so we're insulated from unexpected changes, but I would at least like to know if something breaks. Mapping the servers to some other format seems (to me) like extra work for not a whole lot of gain, since we'd have to validate:

- our string mapping retains the same order as the bottle server mapping
- all of the bottle adapters are mapped to string representations
- we have no extra string representations that aren't represented in the bottle adapters

which I guess you could verify with something like `tuple(map(to_string_mapping, bottle.adapters)) == tuple(string_mapping_to_import_check_map)`.

We could do that in unit tests, which does reduce coupling to bottle at the application layer, and so is probably not a bad idea, but the coupling still exists anyway and remains intimately tied to bottle's adapter protocol. 

I'll play with it a bit and see what it feels like. If nothing else, I'll throw a TODO in to improve the decoupling in a future commit

Reply:
something like this?

somewhere in `__main__.py`

```python
  server_methods = [('waitress', 'waitress'), ('paste', 'paste'), ('twisted', 'twisted.web'), ('cherrypy', 'cherrypy.wsgiserver'), ('wsgiref','wsgiref')]
```

somewhere in the tests:
```python
import pypiserver.bottle
def test_bottle_auto_resolution_order():
    adapter_rmap = {
        val: key for key, val in pypiserver.bottle.server_names.items()
    }
    assert [adapter_rmap[a] for a in pypiserver.bottle.AutoServer.adapters] == [
        s[0] for s in server_methods
    ]

Reply:
Okay I added this (using an enum instead of strings) in a87e771. The tests wound up being a bit awkward because we need to do two separate import checks for cherry py, so it required deduping, but I think it's reasonable.

Reply:
Hmm I don't quite understand the benefit of using Enum in this case. I get the idea, but the result doesn't look necessarily nice imho. But sure :)

I think you might be able to reduce the awkwardness by changing the signature of `_can_import` to `def _can_import(*names: str)` (add a for loop in body), let the generator expr be `(s for s, *i in AUTO_SERVER_IMPORTS if _can_import(*i))` and the IMPORT `(AutoServer.CherryPy, "cheroot.wsgi", "cherrypy.wsgiserver")` so you'd have unique entries in there

The assert statements seem like they can then be replaced by a simple `==` assertion on the AUTO_SERVER_IMPORTS first items lower str

Reply:
For typing, I prefer having a defined enumeration of options than a general `str`, since the variants of a string are unlimited. Returning an enum allows callers to potentially verify statically that they've covered all the cases. I think the code itself is fine enough. The tests are a bit awkward, but nothing too crazy.

The asserts can't be replaced by a regular `==` because the enumeration names don't exactly match the `ServerAdapter` class names, by virtue of wanting to decouple them.